### DISCUSSION TITLE
Why Shell Scripts

### DISCUSSION THREAD

**Discussion by seandenigris** [hpi-swa/smalltalkCI#590] [General]
As I flop around trying to learn Bash in order to add GToolkit as a supported platform, I am left wondering: why not have a thin layer of shell scripts in order to get a minimal Smalltalk image as soon as possible in which to continue the setup work as opposed to doing everything in shell scripts? Has this been contemplated? Worked on?

Also, how are all the shell scripts tested? I'd like to write tests for the GToolkit stuff but I'm not sure where to look. Upon further investigation, I see [the tests folder](https://github.com/hpi-swa/smalltalkCI/tree/master/tests). How does one run the tests locally? Is there documentation? Also, the prospect of having to understand this new ecosystem brings me back to my previous question: why not test in Smalltalk?

**Comment by gcotelli** [comment]
For testing the bash scripts, take a look at https://github.com/hpi-swa/smalltalkCI/tree/master/tests . It's using shunit2 .

**Comment by seandenigris** [comment]
Thanks, that helps. I see shunit2 seems to have decent [docs](https://github.com/kward/shunit2/blob/master/README.md).

Although, why not use something in Smalltalk like this rather than spend more time in the horrifying world of shell scripts:
<img width="1926" alt="Screenshot 2023-03-01 at 11 59 07 AM" src="https://user-images.githubusercontent.com/184176/222209620-a7e55f80-2866-49b9-98aa-b45a615a9954.png">

I hope it doesn't sound like I'm complaining. I'm just wondering if this has ever been considered and maybe even rejected for a good reason I'm not seeing...

**Comment by fniephaus** [comment]
> I'm just wondering if this has ever been considered and maybe even rejected for a good reason I'm not seeing...

SmalltalkCI tries to do as little as possible in Bash for bootstrapping, but that still is more than we like. We once considered moving more setup bits to Smalltalk, but that would've meant that you need to download another image and another VM, and that only makes the service slower and increases the potential for errors (e.g., bad internet connection).
For the same reason, smalltalkCI tries to keep its dependencies to an absolute minimum, which is why we can't use something like `jq` for JSON processing (although [we do use `jq`](https://github.com/hpi-swa/smalltalkCI/blob/5642d387607b29c144aa0b1fc3c4a61bd2894c52/helpers.sh#L387) to report coverage).

If you have any ideas how to improve the bootstrapping process, please let us know or open PRs. :)

**Comment by seandenigris** [comment]
Please have a look at [this spike](https://github.com/seandenigris/Dynabook/tree/master/src/Dynabook-SmalltalkCI) which allows unit testing shell scripts from within Smalltalk. That's how I tested the GToolkit platform support code. Let me know if you think it is worthwhile pursuing.

<img width="2332" alt="Screenshot 2023-03-01 at 9 40 29 PM" src="https://user-images.githubusercontent.com/184176/222317129-bf9bb95f-712e-4889-8f3a-24a1a9ebc1ca.png">

**Comment by fniephaus** [comment]
Testing Bash code in Smalltalk is of course an option, but may also be confusing and not easy to discover because you typically edit Bash and Smalltalk code in different IDEs (not sure you're editing Bash in GToolkit as well). I'd be inclined to keep it simple and just add some simple Bash tests for the GToolkit integration.

**Reply by seandenigris** [reply]
Thanks for the reply. I wanted to edit the scripts in GT, but that will be a project for another day. While "editing" is easy enough as raw text, the value added would be to host bash as a supported language to use e.g. the browser, senders, refactoring. This was way beyond the amount of time I had available. Yes, I guess conforming to the existing infrastructure would be easiest.

**Comment by seandenigris** [comment]
Thanks everyone for the dialogue. It was illuminating
